docs/design/2022-06-15-phyical_plan_explain.md

```diff
@@ -0,0 +1,603 @@
+     `c` varchar(255) DEFAULT NULL
+   ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin"
+   ```
+
+   ```sql
+   SELECT * FROM t2 where a>1
+   ```
+
+   ```Text
+   == Physical Plan ==
+   *(1) ColumnarToRow
+   +- TiKV CoprocessorRDD{[table: t1] TableScan, Columns: a@UNSIGNED LONG, Residual Filter: [a@UNSIGNED LONG GREATER_THAN 1], PushDown Filter: [a@UNSIGNED LONG GREATER_THAN 1], KeyRange: [([t\200\000\000\000\000\000\004W_r\000\000\000\000\000\000\000\000], [t\200\000\000\000\000\000\004W_s\000\000\000\000\000\000\000\000])], Aggregates: , startTs: 433780573880188929}
+   ```
+
+2. unclear representation of the execution process
+
+   For an IndexScan, normally the IndexRangeScan should be executed first, then the TableRowIDScan. Only in the case of triggering a downgrade will the Downgrade Filter be executed, but in the physical execution plan, only the Downgrade Filter is shown.
+
+   ```SQL
+   CREATE TABLE `t2` (
+     `a` BIGINT(20) UNSIGNED  NOT NULL,
+     `b` varchar(255) NOT NULL,
+     `c` varchar(255) DEFAULT NULL,
+     PRIMARY KEY (`a`,`b`)
+   ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
+   ```
+
+   ```SQL
+   SELECT * FROM t2 where a<1 or b>'bb'"
+   ```
+
+   ```text
+   == Physical Plan ==
+   *(1) ColumnarToRow
+   +- TiSpark RegionTaskExec{downgradeThreshold=1000000000,downgradeFilter=[[[a@UNSIGNED LONG LESS_THAN 1] OR [b@VARCHAR(255) GREATER_THAN "bb"]]]
+      +- RowToColumnar
+         +- TiKV FetchHandleRDD{[table: t2] IndexScan[Index: primary] , Columns: a@UNSIGNED LONG, b@VARCHAR(255), c@VARCHAR(255), Downgrade Filter: [[a@UNSIGNED LONG LESS_THAN 1] OR [b@VARCHAR(255) GREATER_THAN "bb"]], PushDown Filter: [[a@UNSIGNED LONG LESS_THAN 1] OR [b@VARCHAR(255) GREATER_THAN "bb"]], KeyRange: [([t\200\000\000\000\000\000\003\374_i\200\000\000\000\000\000\000\001\000], [t\200\000\000\000\000\000\003\374_i\200\000\000\000\000\000\000\001\372])], Aggregates: , startTs: 433735624236728323}
+   ```
+
+3. obscure push-down conditions
+
+   As shown below, for the query condition a>0 is pushed down, but the pushed-down information is only implicitly given inside the KeyRange, which is not convenient for users to understand.
+
+   ```SQL
+   CREATE TABLE `t3` (
+     `a` BIGINT(20) NOT NULL,
+     `b` varchar(255) NOT NULL,
+     `c` varchar(255) DEFAULT NULL,
+     PRIMARY KEY (`a`) clustered
+   ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
+   ```
+
+   ```SQL
+   SELECT a FROM t3 where a>0
+   ```
+
+   ```text
+   == Physical Plan ==
+   *(1) ColumnarToRow
+   +- TiKV CoprocessorRDD{[table: t2] CoveringIndexScan[Index: primary(a,b)] , Columns: a@UNSIGNED LONG, b@VARCHAR(255), Residual Filter: [b@VARCHAR(255) GREATER_THAN "aa"], PushDown Filter: [b@VARCHAR(255) GREATER_THAN "aa"], KeyRange: [([t\200\000\000\000\000\000\005z_i\200\000\000\000\000\000\000\001\000], [t\200\000\000\000\000\000\005z_i\200\000\000\000\000\000\000\001\372])], Aggregates: , startTs: 433789855560368130}
```

Comment:
Is this the right plan displayed? Why there is a `Residual Filter: [b@VARCHAR(255) GREATER_THAN "aa"], PushDown Filter: [b@VARCHAR(255) GREATER_THAN "aa"]`?  And table is `t2`?

Reply:
It is old, I forget to replace it.